### ISSUE TITLE
question of bert model size

### ISSUE THREAD

**Issue by miangangzhen** [ymcui/Chinese-BERT-wwm#22]
Just wonder that how to get a smaller bert model like this one? Normal bert model is 1G, but this one is only 400M.

Thank you very much.

**Comment by ymcui** [comment]
Hi,
As Adam-like optimizer will save additional params, the resulting model will be about 3x larger.
If you would like to shrink your model size, a straightforward way would be using 'tf.train.GradientDescentOptimizer' (change this line: https://github.com/google-research/bert/blob/master/optimization.py#L59) and initialize with your desired model till the initial checkpoint `model.ckpt-0.*` is saved (better to save in another folder).
Then you will find `model.ckpt-0.*` is about 1/3 size of the original one (w/o additional param for optimizer).
It is also recommended to verify this model on your data to ensure it is exactly the same with the original one.

**Comment by ymcui** [comment]
reopen if necessary

**Comment by stevewyl** [comment]
I do not understand "initialize with your desired model till the initial checkpoint model.ckpt-0.* is saved". After loading checkpoint, optimizer only initialized in TRAIN MODE. So I need to create a new `train_op` that use SGD but actually not train the model anymore. Is that correct? Is there any more straightforward way? Load model, remove useless nodes and then save the new model.

**Comment by ymcui** [comment]
1. 在optimizer.py里把AdamW改为SGD
2. 运行run_pretraining.py，指定init_checkpoint为你要缩减大小的模型
3. 待显示完'Saving checkpoints for 0 into xxxxx' 你要的缩减模型就保存在xxx路径了

已经很直白了，基本不需要你改多少代码。

**Comment by stevewyl** [comment]
@ymcui 感谢！可以了👍
